File: decorator/java/Sauce.java
package decorator.java;

public class Sauce implements Food {
    private final Food food;

    public Sauce(Food food) {
        this.food = food;
    }

    @Override
    public void dip() {
        food.dip();
        System.out.println("Adding sauce to the food");
    }
}

File: decorator/golang/sauce.go
package main

type Sauce struct {
	Food Food
}

func NewSauce(food Food) *Sauce {
	return &Sauce{
		Food: food,
	}
}

func (s *Sauce) dip() {
	println("Adding sauce to the food and dipping it")
	s.Food.dip()
}
